### DISCUSSION TITLE
How to get a customize tar.xz release

### DISCUSSION THREAD

**Discussion by firedent** [transmission/transmission#2746] [Q&A] [answered]
Hey all,

I was working on my own customized TR recently and building a DSM package via [spksrc](https://github.com/firedent/spksrc). I notice that the spksrc will download the sources(transmission-3.00.tar.xz) from [releases page](https://github.com/transmission/transmission/releases) and cross-build those sources. I find that the content of transmission-3.00.tar.xz is different from the [source tar file](https://github.com/transmission/transmission/archive/refs/heads/main.zip) downloading from the repo source page. 

So, how could I fork TR, modify it and release a new transmission-3.00.tar.xz which has the same content as the official repo on my forked repo? After reading the README, WIKI and automaker manual, I still can't find the solution. 

Kind Regards,
Shichao

**Comment by mikedld** [comment, marked as answer]
One obvious difference would be that archive downloaded directly from GitHub will not have files provided via submodules. If you want to create a tarball the way we do it, I suggest starting by looking through the commands executed on CI, and maybe simplifying them a bit if desired. E.g. in https://build.transmissionbt.com/job/trunk-linux/10107/console you'll see
```
...
(checkout and update submodules recursively, as described README.md)
...
06:04:14 + cmake -S . -B _tarball -G Ninja -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCPACK_SOURCE_IGNORE_FILES=_tarball -DENABLE_CLI=ON
...
06:04:40 + cmake --build _tarball --target package_source
...
```

**Comment by firedent** [comment]
@mikedld Thank you for your reply.

After reading your comment, I was building my own CI env. Could you please specify the CMake version?

I got a Segmentation fault with CMake 3.16.3 provided by Ubuntu official repo.
```
[0/1] Run CPack packaging tool for source...
CPack: Create package using TXZ
CPack: Install projects
CPack: - Install scripts: /home/ubuntu/transmission/cmake/CPackSourceFixDirLinks.cmake
CPack: - Install script: /home/ubuntu/transmission/cmake/CPackSourceFixDirLinks.cmake
-- Fixing link: macosx/Sparkle.framework/Versions/Current -> A
-- Fixing link: macosx/Sparkle.framework/Resources -> Versions/Current/Resources
-- Fixing link: macosx/Sparkle.framework/PrivateHeaders -> Versions/Current/PrivateHeaders
-- Fixing link: macosx/Sparkle.framework/Modules -> Versions/Current/Modules
-- Fixing link: macosx/Sparkle.framework/Headers -> Versions/Current/Headers
CPack: - Install directory: /home/ubuntu/transmission
Segmentation fault (core dumped)
FAILED: CMakeFiles/package_source.util
cd /home/ubuntu/transmission/_tarball && /usr/bin/cpack --config ./CPackSourceConfig.cmake /home/ubuntu/transmission/_tarball/CPackSourceConfig.cmake
ninja: build stopped: subcommand failed.
```

When I use the latest CMake (3.23.0-rc2) installed via official .sh, I got 'File exists' error.
```
[0/1] Run CPack packaging tool for source...
CPack: Create package using TXZ
CPack: Install projects
CPack: - Install scripts: /home/ubuntu/transmission/cmake/CPackSourceFixDirLinks.cmake
CPack: - Install script: /home/ubuntu/transmission/cmake/CPackSourceFixDirLinks.cmake
-- Fixing link: macosx/Sparkle.framework/Versions/Current -> A
-- Fixing link: macosx/Sparkle.framework/Resources -> Versions/Current/Resources
-- Fixing link: macosx/Sparkle.framework/PrivateHeaders -> Versions/Current/PrivateHeaders
-- Fixing link: macosx/Sparkle.framework/Modules -> Versions/Current/Modules
-- Fixing link: macosx/Sparkle.framework/Headers -> Versions/Current/Headers
CPack: - Install directory: /home/ubuntu/transmission
CMake Error: failed to create symbolic link 'macosx/Sparkle.framework/Resources': File exists
CPack Error: Cannot create symlink: macosx/Sparkle.framework/Resources--> Versions/Current/Resources
CPack Error: Error when generating package: transmission
FAILED: CMakeFiles/package_source.util
cd /home/ubuntu/transmission/_tarball && /home/ubuntu/cmake-3.23.0-rc2-linux-x86_64/bin/cpack --config ./CPackSourceConfig.cmake /home/ubuntu/transmission/_tarball/CPackSourceConfig.cmake
ninja: build stopped: subcommand failed.
```
The Segmentation fault disappeared. So, I think maybe I should use the specific version of CMake.

**Comment by mikedld** [comment]
The code that fixes up symlinks was added to fix an issue with CPack that wasn't preserving them for some reason, then leading to broken source tarballs. Using CMake 3.22.2 locally I now see that this is no longer the case, not sure when this happened. Jenkins builds are running on Fedora 34 and using CMake version that comes with it, i.e. 3.20.5. TeamCity builds are running on Gentoo and currently using CMake 3.18.2 and 3.19.3.

**Reply by firedent** [reply]
Thank you for your information. I tested a variety of CMake and distribution combinations for excluding that is caused by my incorrect installation.

Ubuntu 20 with CMake 3.23.0-rc2 binary install  on AWS: NO
Ubuntu 20 with CMake 3.22.3 binary install  on AWS: NO
Ubuntu 20 with CMake 3.22.3 on GitHub Action: NO
Ubuntu 20 with CMake 3.22.2 binary install on AWS: NO
Ubuntu 20 with CMake 3.22.2 source install on AWS: NO

Fedora 35 with CMake 3.22.2 come with official repo on Docker: NO
Fedora 34 with CMake 3.20.5 come with official repo on Docker: YES
Red Hat 8 with CMake 3.20.2 come with official repo on AWS: YES

In the beginning, I thought picking the right CMake is all, but I'm wrong. Finally, I chose Fedora 34 with CMake 3.20.5

BTW, according to the [CMakeLists](https://github.com/transmission/transmission/blob/bb6b5a062ee594dfd4b7a12a6b6e860c43849bfd/CMakeLists.txt#L568) of 3.00 release, there should be a tar.bz2 file instead of [transmission-3.00.tar.xz](https://github.com/transmission/transmission/releases/download/3.00/transmission-3.00.tar.xz). So, you just converted it on CI. Am I right?

**Comment by ckerr** [comment]
This ticket isn't so much about a Transmission bug as a question, so I'm going to convert this issue into a discussion topic

**Comment by mikedld** [comment]
I'll look into those symlinks once again, something's definitely wrong.

As for the archive format, 3.00 tarball was prepared by autotools since we didn't fully switch to CMake back then yet:
https://github.com/transmission/transmission/blob/bb6b5a062ee594dfd4b7a12a6b6e860c43849bfd/configure.ac#L19

Later on we removed autotools build system, left CMake only and fixed its configuration up to produce .tar.xz which is what you could see in `main` right now (and what will be there for 4.x release): https://github.com/transmission/transmission/blob/7a227917ff3e9fc15cc49164d8801774f10593e1/CMakeLists.txt#L688
